### DISCUSSION TITLE
question about understanding floem

### DISCUSSION THREAD

**Discussion by sbechet** [lapce/floem#337] [Q&A]
I'm in the process of discovering lapce and this project is fantastic. But i have a beginner's question, please don't take my question the wrong way: I am not an expert in graphical interface. Can you help me understand what floem offers compared to egui? egui seems older with lots of widgets and seems to refresh only as needed like floem (optimization), so I'm having trouble understanding the purpose.

**Comment by sbechet** [comment]
Maybe because it implements ropescience?

**Reply by MinusGix** [reply]
Ropes are just for the editor logic itself, the tl;dr is that it makes certain operations faster than having a `String`

**Comment by jrmoulton** [comment]
I'm writing as a user of floem and one who has never used egui. 

egui is an immediate mode gui library. This means that on every frame egui will re-render everything in the app. This makes it so that elements respond to changes in state because every value is checked on every frame. I don't understand all of the reasons why but this also makes it so that egui can't use more complex layout algorithms so layouts in egui are very simple and relatively restricted but it is still very functional. 

Floem uses "signals" to manage reactivity. Signals have become popular with many JS frameworks because they offer the least overhead for reactive state (high performance) while also being very very easy to use.

For me floem is great because of the high performance and light weight resource usage. It is great because the components are declarative and composable. The reactive styling system is also incredible (it is really really good). 

As a user to sum it up: Performance and the best developer ergonomics.

**Reply by sbechet** [reply]
Thanks. Strangely, `egui` doesn't update anything (no code calls) if there is nothing to update on screen. In my [example](https://codeberg.org/sbechet/dodometer/src/branch/main/crates/starfield/) (see `examples/starfield.rs` line 19), I am forced to call `ui.ctx().request_repaint();` if I want to refresh forcefully at each update. There must still be some subtlety that I haven't understood between floem and egui. Maybe one day...

**Comment by NoelJacob** [comment]
Immediate mode is the coding pattern for the GUI and only optimized calls are given to the GPU to redraw. https://github.com/ocornut/imgui#how-it-works Explanation of immediate mode for another well known library.

**Reply by sbechet** [reply]
Thank you.
